CRM/Core/Payment.php

```diff
@@ -1953,4 +1953,114 @@ public function isSendReceiptForPending() {
     return FALSE;
   }
 
+  /**
+   * doCheckout is an alternative to doPayment used to process afform_payments
+   *
+   * The base implementation just piggybacks on doPayment, but different payment processors
+   * will need to implement differently depending on the user flow through checkout
+   *
+   * NOTE: no Payment is created in CiviCRM in this implmentation
+   *  - for some processors this will be done later, in response to a postback from the payment
+   *    processor's infrastructure confirming the payment has been made
+   *  - some processors may want to complete the payment within doCheckout
+   *
+   * @param Civi\Payment\PropertyBag $paymentParams
+   *   params for the payment - as would be passed to doPayment. this is most commonly an array
+   *   of line items for the contribution, and some details about the contact doing the paying
+   * @param ?string $successUrl
+   *   url to take the user to after successful payment
+   * @param ?string $failUrl
+   *   url to take the user to if payment is unsuccessful
+   *
+   * @return array
+   *   response indicating the next step for the user. what form this takes is WIP,
+   *   but currently this is consumed by Contribution.pay
+   *
+   * @see \Civi\Api4\Action\Contribution\Pay
+   *  (in afform_payment extension)
+   */
+  public function doCheckout(PropertyBag &$paymentParams, ?string $successUrl, ?string $failUrl): array {
```

Comment:
So @mattwire suggested we look at the signatures here - my hot takes are 

1) if we are going to use `PropertyBag` I think we need to address the usability - the primary issue is the fact that you have to use `has` before `get` because the type hints are such that it errors when null

2) I think we should pass the urls as an array

3) The doPayment function just returns when successful & otherwise the PaymentProcessorException is thrown - is there a reason not to follow that? Does it cause issues in Afform?

Reply:
Ive changed this to `array`